### DISCUSSION TITLE
rearrange edge_index

### DISCUSSION THREAD

**Discussion by Chrisable** [pyg-team/pytorch_geometric#5214] [General]
How to get rearranged edge_index?
my edge_index:
```
tensor([[  0,   0,   0,  ..., 204, 204, 204],
        [  0,  37,  42,  ..., 146, 197, 204]])
```
But it only has 106 nodes,how to Rearrange edge_index so that its value range is:[0,105]
for example:
```
tensor([[  0,   0,   0,  ..., 105, 105, 105],
        [  0,  10,  37,  ...,  74, 100, 105]])
```

**Comment by EdisonLeeeee** [comment]
Hi. Maybe you can take a look at `remove_isolated_nodes`:

```python
>>> edge_index = torch.LongTensor([[0, 100], [100, 0]])
>>> edge_index
tensor([[  0, 100],
        [100,   0]])

>>> from torch_geometric.utils import remove_isolated_nodes
>>> edge_index = remove_isolated_nodes(edge_index)[0]
>>> edge_index
tensor([[0, 1],
        [1, 0]])
```

**Reply by Chrisable** [reply]
Thank you for your reply.It helps a lot

**Reply by Chrisable** [reply]
But i get a unexpected result,
my edge_index:
```
tensor([[  0,   0,   0,  ..., 204, 204, 204],
        [  0,  37,  42,  ..., 146, 197, 204]])
```
result:
```
tensor([[  0,   0,   0,  ..., 103, 104, 105],
        [ 16,  18,  27,  ..., 103, 104, 105]])
```
And the last 105 columns of the results are:
```
[[...,0,   1,   2,   3,   4,   5,   6,   7,   8,
          9,  10,  11,  12,  13,  14,  15,  16,  17,  18,  19,  20,  21,
         22,  23,  24,  25,  26,  27,  28,  29,  30,  31,  32,  33,  34,
         35,  36,  37,  38,  39,  40,  41,  42,  43,  44,  45,  46,  47,
         48,  49,  50,  51,  52,  53,  54,  55,  56,  57,  58,  59,  60,
         61,  62,  63,  64,  65,  66,  67,  68,  69,  70,  71,  72,  73,
         74,  75,  76,  77,  78,  79,  80,  81,  82,  83,  84,  85,  86,
         87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,  98,  99,
        100, 101, 102, 103, 104, 105]
[...,0,   1,   2,   3,   4,   5,   6,   7,   8,
          9,  10,  11,  12,  13,  14,  15,  16,  17,  18,  19,  20,  21,
         22,  23,  24,  25,  26,  27,  28,  29,  30,  31,  32,  33,  34,
         35,  36,  37,  38,  39,  40,  41,  42,  43,  44,  45,  46,  47,
         48,  49,  50,  51,  52,  53,  54,  55,  56,  57,  58,  59,  60,
         61,  62,  63,  64,  65,  66,  67,  68,  69,  70,  71,  72,  73,
         74,  75,  76,  77,  78,  79,  80,  81,  82,  83,  84,  85,  86,
         87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,  98,  99,
        100, 101, 102, 103, 104, 105]
]
```

**Comment by Chrisable** [comment]
it is ok.edge_index exists some self-loop

**Comment by Chrisable** [comment]
By the way,how to get original edge_index?

**Comment by EdisonLeeeee** [comment]
`remove_isolated_nodes` returns a node mask, where the preserved node are marked as `True`.
```python
edge_index_remain, _, mask = remove_isolated_nodes(edge_index)
edge_mask = torch.logical_and(mask[edge_index[0]], mask[edge_index[1]])
edge_index[:, edge_mask]
```

**Reply by Chrisable** [reply]
Thank for very much!

**Comment by Chrisable** [comment]
Since the characteristics of isolated nodes are still helpful for GNN training, can we rearrange the nodes while preserving the self loop?

**Reply by rusty1s** [reply]
Can you clarify what you mean?

**Reply by Chrisable** [reply]
Remove_isolated_nodes() means that the isolation nodes will be removed first, and then for edge_ Index rearranges the nodes. If I only want to rearrange the nodes and do not want to remove the isolation nodes, what should I do?

**Reply by rusty1s** [reply]
I am not sure I understand. If you only want to rearange the indices in edge indices without modifying the order in the node feature matrix, this will lead to wrong results.

**Reply by Chrisable** [reply]
I mean that 

my edge_index:

```
tensor([[  0,   0,   0,  ..., 204, 204, 204],
        [  0,  37,  42,  ..., 146, 197, 204]])
```
But it only has 106 nodes,how to rearrange edge_index and do not delete the isolated nodes so that its value range is:[0,105] 
for example:

tensor([[  0,   0,   0,  ..., 105, 105, 105],
        [  0,  10,  37,  ...,  74, 100, 105]])

**Comment by EdisonLeeeee** [comment]
```python
import torch
edge_index = torch.tensor([[100, 200, 300, 400], [200, 300, 400, 500]])
>>> edge_index
tensor([[100, 200, 300, 400],
        [200, 300, 400, 500]])

>>> _, edge_index = torch.unique(edge_index, return_inverse=True)
>>> edge_index
tensor([[0, 1, 2, 3],
        [1, 2, 3, 4]])
```
Is that what you want?

**Reply by Chrisable** [reply]
yes,thank you for your help.
